### PR TITLE
[libc][math] Implement nexttoward functions

### PR DESCRIPTION
**Pull request by nishantwrp**
Implements the `nexttoward`, `nexttowardf` and `nexttowardl` functions. Also, raise excepts required by the standard in `nextafter` functions. 

cc: @lntue

### CODE REVIEW

File: libc/src/math/generic/nexttowardl.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#include "src/__support/FPUtil/ManipulationFunctions.h"
+#include "src/__support/common.h"
+
+namespace LIBC_NAMESPACE {
+
+LLVM_LIBC_FUNCTION(long double, nexttowardl, (long double x, long double y)) {
+  // We can reuse the nextafter implementation because nexttoward behaves
+  // exactly same as nextafter in case of long doubles. Also, we have explcitly
+  // handled the special 80-bit long doubles in nextafter implementation.
+  return fputil::nextafter(x, y);
```
Comment (nishantwrp) [inline libc/src/math/generic/nexttowardl.cpp:19]: Is this the correct place to reuse the `nextafter` implementation? Or should I create a new function in FpUtils for `nexttoward` and call this function there.

File: libc/src/__support/FPUtil/ManipulationFunctions.h [hunk 2]
```diff
@@ -169,8 +171,49 @@ LIBC_INLINE T nextafter(T from, T to) {
     int_val = (to_bits.uintval() & sign_mask) + UIntType(1);
   }
 
+  UIntType exponent_bits = int_val & FloatProperties<T>::EXPONENT_MASK;
+  if (exponent_bits == UIntType(0))
+    raise_except_if_required(FE_UNDERFLOW | FE_INEXACT);
+  else if (exponent_bits == FloatProperties<T>::EXPONENT_MASK)
+    raise_except_if_required(FE_OVERFLOW | FE_INEXACT);
```
Comment (nishantwrp) [inline libc/src/__support/FPUtil/ManipulationFunctions.h:178]: Should probably test the exceptions in the tests for nextafter too. Will do that in a separate pr?

File: libc/src/__support/FPUtil/ManipulationFunctions.h [hunk 3]
```diff
@@ -169,8 +171,49 @@ LIBC_INLINE T nextafter(T from, T to) {
...
+
+  return cpp::bit_cast<T>(int_val);
+}
+
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE T nexttoward(T from, long double to) {
```
Comment (lntue) [inline libc/src/__support/FPUtil/ManipulationFunctions.h:184]: For free function, I personally prefer to use SFINAE on the returning type `LIBC_INLINE cpp::enable_if_t<cpp::is_floating_point_v<T>, T> nexttoward(...)`, so that it frees the second template parameter to be used if needed.

File: libc/src/__support/FPUtil/ManipulationFunctions.h (hunk 2 above)
Comment (lntue) [reply libc/src/__support/FPUtil/ManipulationFunctions.h:178]: SGTM.  Thanks for fixing this TODO item.

File: libc/src/__support/FPUtil/ManipulationFunctions.h [hunk 4]
```diff
@@ -19,6 +20,7 @@
 #include "src/__support/macros/attributes.h"
 #include "src/__support/macros/optimization.h" // LIBC_UNLIKELY
 
+#include <fenv.h>
```
Comment (lntue) [inline libc/src/__support/FPUtil/ManipulationFunctions.h:23]: You should include `FEnvImpl.h` instead.  It is where `raise_except_if_required` is defined.

File: libc/src/math/generic/CMakeLists.txt [hunk 5]
```diff
@@ -1474,6 +1474,42 @@ add_entrypoint_object(
...
+add_entrypoint_object(
+  nexttoward
+  SRCS
+    nexttoward.cpp
+  HDRS
+    ../nexttoward.h
+  DEPENDS
+    libc.src.__support.FPUtil.manipulation_functions
+  COMPILE_OPTIONS
+    -O2
```
Comment (lntue) [inline libc/src/math/generic/CMakeLists.txt:1486]: I think it is ok to change all of these to `-O3`.

File: libc/src/math/generic/nexttowardl.cpp (hunk 1 above)
Comment (lntue) [reply libc/src/math/generic/nexttowardl.cpp:19]: It is ok to re-use any `fputil` functions.  We just try not to call other functions that are implemented under `LLVM_LIBC_FUNCTION`, i.e. other entry point targets.

**Review by lntue — COMMENTED** [review]
Thanks for adding these functions!  Overall LGTM with a few nits.  Do you mind also updating the followings:
- bazel build: https://github.com/llvm/llvm-project/blob/main/utils/bazel/llvm-project-overlay/libc/BUILD.bazel
- progress documentation: https://github.com/llvm/llvm-project/blob/main/libc/docs/math/index.rst

Thanks!

File: libc/src/__support/FPUtil/ManipulationFunctions.h (hunk 3 above)
Comment (lntue) [reply libc/src/__support/FPUtil/ManipulationFunctions.h]: Essentially, when we do SFINAE with `template<T, U = enable_if_t<..., void>>`, we do not intend to let user use the second template parameter, but nothing prevents user code to be `func<T, U>(...)`, and it might silently get wrong function.  If all users do is to call `func(...)` or `func<T>` then it doesn't matter.

**Review by lntue — APPROVED** [review]

File: libc/src/__support/FPUtil/ManipulationFunctions.h [hunk 6]
```diff
@@ -169,8 +171,50 @@ LIBC_INLINE T nextafter(T from, T to) {
...
+template <typename T>
+LIBC_INLINE cpp::enable_if_t<cpp::is_floating_point_v<T>, T>
+nexttoward(T from, long double to) {
+  FPBits<T> from_bits(from);
+  if (from_bits.is_nan())
+    return from;
+
+  FPBits<long double> to_bits(to);
+  if (to_bits.is_nan())
+    return to;
```
Comment (nickdesaulniers) [inline libc/src/__support/FPUtil/ManipulationFunctions.h:192]: I'm observing the following warnings when building `ninja check-libc` with clang:
```
[286/3491] Building CXX object projects/libc/src/math/gene...es/libc.src.math.generic.nexttowardf.dir/nexttowardf.cpp.o
In file included from /android0/llvm-project/libc/src/math/generic/nexttowardf.cpp:10:
/android0/llvm-project/libc/src/__support/FPUtil/ManipulationFunctions.h:191:12: warning: implicit conversion loses floating-point precision: 'long double' to 'cpp::enable_if_t<cpp::is_floating_point_v<float>, float>' (aka 'float') [-Wimplicit-float-conversion]
    return to;
    ~~~~~~ ^~
/android0/llvm-project/libc/src/math/generic/nexttowardf.cpp:16:18: note: in instantiation of function template specialization '__llvm_libc_18_0_0_git::fputil::nexttoward<float>' requested here
  return fputil::nexttoward(x, y);
                 ^
In file included from /android0/llvm-project/libc/src/math/generic/nexttowardf.cpp:10:
/android0/llvm-project/libc/src/__support/FPUtil/ManipulationFunctions.h:194:12: warning: implicit conversion loses floating-point precision: 'long double' to 'cpp::enable_if_t<cpp::is_floating_point_v<float>, float>' (aka 'float') [-Wimplicit-float-conversion]
    return to;
    ~~~~~~ ^~
```

Comment (michaelrj-google) [reply libc/src/__support/FPUtil/ManipulationFunctions.h:192]: That issue should be fixed by https://github.com/llvm/llvm-project/pull/73698. Are you on an updated checkout?

Comment (nickdesaulniers) [reply libc/src/__support/FPUtil/ManipulationFunctions.h:192]: ah, thanks, slightly behind!
